### ISSUE TITLE
[MOD = 3.6][MC = 1.11.2][Questions about pre-generation]

### ISSUE THREAD

**Issue by Endmateria** [ternsip/StructPro#18]
Sorry, I know you have written extensive details on the usage of this mod, but there are still a couple things I'm curious about regarding the mechanics of this mod.
1.  Does anything undesirable happen if you use the pre-generator on the same chunks multiple times?  Like does it register that it's already added buildings in that area, or does it ignore that fact and try to place more?
2.  I'm getting a moderate amount of cascading generation lag (in pre-generation).  Does your pre-generator normally cause that, or is that more likely caused by mods I have installed that may have poor (runaway) generation code?  If the latter is true then do you know which types of terrain generation assets are the usual culprits?  Mods that add large batches of assets that overlap?  I mean, I've got the Coral reef mod that spawns in glowing coral in patches, Quark, which manipulates the placement of underground stone and adds huge underground caves/biomes.  (I'm willing to bet it's one of those two).  Better Bedrock Gen, which makes all the bedrock flat.  BOP which does a lot in and of itself.  And then the rest are just mods that add small amounts of ore here and there (probably not the cause).  Or could it be that I simply have too much crap generating? hahahaha
3. If you start the pre-generator and select a size without selecting a starting point, where does it start?

**Comment by ternsip** [comment]
1. Of course it does not. It generates one time per one chunk population. All structures that started in specific chunk will be spawned during chunk generation. Each chunk generation always happens one time. So you can pre-generate multiple times, but any chunk in such range has already been generated it just loads into memory, not generate it again.
2. Go to server.properties and set max-tick-time=-1. It happens because of structpro because pasting large structures cause processing of falling gravel, sand, water, lava, mobs etc. But I'm not totally sure that everything is absolutely correct.
3.  It starts in (0, 0) chunk

**Comment by Endmateria** [comment]
Yeah, I read just about everything in your information section.  Got the max tickrate set at -1 already... that's why I asked if it was being caused by other mods.  Might just be the amount of mods I've got generating stuff...  https://www.reddit.com/r/feedthebeast/comments/5x0twz/investigating_extreme_worldgen_lag/

**Comment by ternsip** [comment]
I have to say that I additionally save and unload all redudant chunks each time I generate or load it to the memory, so I believe the process can be a lil bit optimized by checking if chunk has already been generated and skip it. But the main problem is why genberation start stuttering is opened.

**Comment by Endmateria** [comment]
Given how long the pre-generation process can take, having an option to "skip already generated chunks" would be great!  To your other question, I don't know, but I'm getting cascading generation, and there's a good chance that it's not your mod, but somebody else's that isn't optimised.  I might have to slowly take off mods and play with it to see if I can find the culprit.  Your mod is brilliant by the way, not being critical here at all... just want to get it working right while taking off the least amount of other mods as possible haha.

**Comment by ternsip** [comment]
I checked out one of my previous 4097x4097 blocks (size =128 chunks radius) world pre-generation and what I realize:
1. Cascading generation cause some rare third-party mods, because for my mod pack with ic2, biomes o plenty et all was ok and not cascading happened. To reproduce I need list of your mods.
2. Generation consume server tick thread to generate and force populate chunk, so it generates (X,Z), (X+1, Z), (X, Z+1), (X+1,Z+1) blocks immidiatly so no problems caused.
3. Before determination that structure spawns in certain chunk starts surface roughness calibrator to detect roughness, liquid depth, biomes etc to say about spawing permitions. That cause also some chunks ahead to load, but it's not a lot. Big structures like deathstar spawns in 20 seconds that extreemly fast. And if spawn conditions not permit, next time it will recalculates much faster because all in RAM.
4. After some time after pre-generation starts (for me it started on ~63000/66000 chunks progress) Massive lags appears and It's because statistically one of structures with lots of sand and gravel spawned in the sky and start falling down that stutter server for a while.

**Comment by Endmateria** [comment]
Quark and Galacticraft, those are the two culprits I found thus far... I notified them.  Farseek was also posting a lot of activity during the pregen... though I don't know why I even had that mod installed (it doesn't show up as a dependency anywhere).  Let me run some more tests, I'm resorting mods at the moment.  Oh well, here's a mods list (crash report) for shits and giggles...  https://gist.github.com/Endmateria/19b081faa7ec6bfe82cd9c2203b111f4#file-crash-2017-05-06_16-24-58-server-txt

**Comment by ternsip** [comment]
java.lang exceptions happens when you trying to stop server and thread stops asynhronous

**Comment by Endmateria** [comment]
I was in game, the server forcefully stopped itself, or in so many words... crashed.  It's done that multiple times.  I'm of the opinion that your mod, Quark, and Galacticraft don't get along hahaha.

**Comment by Endmateria** [comment]
Here's radfast's response (from Galacticraft)... https://github.com/micdoodle8/Galacticraft/issues/2689#issuecomment-299665741

**Comment by ternsip** [comment]
I read carefully what **radfas** says and I can say that structpro not cause runaway worldgen even when density is crazy overcap, because of next features:
Structpro generates all structures that start in (x,z) chunk that growth only in positive axis directions (+x, +z), so you can see how appears rectangle (x,z) - (x + width, z + length) in the world. And such process not begins to populate border chunks. That means that newly generated chunks fires only direct neighbours to generate (not populate) and the process stop there. To continue the process players move forward and newly generated chunk that is not already populated populates itself and process continues.

**Comment by Endmateria** [comment]
Well, I took off some of the mods that cause cascading world gen... Coral reef and Tinker's construct.  Astral Sorcery has corrected the problem, just waiting for him to release the updated version.  Your mod alone is still getting cascading world gen during pregen though.  I'm running more tests this morning.  Here's a short log excerpt of me pre-generating a 100 chunk radius (40000 chunks) on a new world map.  In this short log, your mod has 18 hits worth of cascading world gen lag.  Could it be the structures you chose, and not necessarily the mod itself?  https://gist.github.com/Endmateria/99b4a59ebcdb10e1c0abcb2195142f4b#file-fml-server-latest-log

**Comment by ternsip** [comment]
I saw your gist and that's totally not related to structpro.

**Comment by ternsip** [comment]
Cascading world gen not happens except rectangle zone where schematics pasts.

**Comment by ternsip** [comment]
![mc-1](https://cloud.githubusercontent.com/assets/10179239/25785198/570fe942-3383-11e7-80ce-9a647bc93160.png)
You can see rectangles on edges that not cascading generation, that's forced generation by mod to check area for spawning conditions as I said before.

**Comment by Endmateria** [comment]
Yeah, this is what I got... similar.  https://ibb.co/mwFtMQ

**Comment by ternsip** [comment]
This small squares is space that schematics required to spawn or to check spawn conditions. Everything is correct. Nothing critical, that's the way how the mod works. There is not other way to implement or improove logic.

**Comment by Endmateria** [comment]
I gotcha, makes sense.  The cascading worldgen was mainly from other mods, whom I've opened up issues with.  My major issue was with the server crashing ingame.... but that only seems to happen when you go near the edges of the world that have not been pre-generated.  Have you considered adding a world borders option to prevent the server from crashing?

**Comment by Endmateria** [comment]
Hey man, saw the options that you added to the /structpro help menu.  Very nice.  And thanks for troubleshooting all of this with me.  It's a very temperamental mod, but once you know all the rules, it works like a charm.  I would recommend specifically warning people on your webpage that the server will crash if you don't pre-gen properly (since crashing seems to be everyone's primary complaint in your issue tracker), and I also recommend telling people to put a world border around the pre-generated area.  Hopefully that will help you avoid having to address all of these concerns for someone else.  Anyways, thanks again... great mod!

**Comment by Endmateria** [comment]
Ok, so the last two times I've run the pre-generator with large sizes (1000+ chunks), I've gotten an "out of memory error" that crashes the server.  It ran for about an hour before crashing.  I've already given the server 8 gigs of ram.  Do I give it more, or is there a memory leak somewhere?  https://gist.github.com/Endmateria/02e99417a3816172e34ab9c89490c021#file-crash-2017-05-09_01-29-27-server-txt

**Comment by Endmateria** [comment]
Hmmm, seems the memory problem could've been a Forge 2294 issue.  I updated to 2296 and I'm not getting out of memory errors anymore, but they've been replaced with "tick out of sync" errors.  Here's two crashes in a row using Forge 2296.  My launch command was /spro gen size=1250 world=0 step=16 world=0 progress=241000 skip=true
 https://gist.github.com/Endmateria/12b502f6011949eb62b8ee1f638d71e7#file-crash-2017-05-09_04-06-22-server-txt
https://gist.github.com/Endmateria/996f471ede9236b36c05ce5c9db6c547#file-crash-2017-05-09_10-32-33-server-txt

And here's the log leading up to the 2nd crash so you can see what was going on...
I did delete some of the OP schematics from the custom village files which is causing a null when those villages search for their pieces.... if that's gonna be a problem then I'll just delete all the custom (Autortown) village schematics.

**Comment by Endmateria** [comment]
Could the memory issue be causing the tickrate desync?  I'm watching the memory allocation for javascript in my task manager and it just keeps growing, meanwhile the generation gets slower and slower (testing at default step=32).  When this crashes, I'll test using step=16 again.

**Comment by Endmateria** [comment]
Yeah, even at step 16 the generator slowly eats up the memory till the generator is crawling (really slow)... then eventually crashes with a "TickNextTick list out of synch" error... https://gist.github.com/Endmateria/442944648b49b3264c369b668ca78228#file-crash-2017-05-09_15-53-18-server-txt

**Comment by Endmateria** [comment]
Also I get this error occasionally while generating using the "progress" feature.  If it's just getting this error because the generator is passing by a structure that it already placed that is no longer loaded into memory, then that makes sense and I'm disregard.  (only happens after using progress/resume).   https://gist.github.com/Endmateria/b5aa25206a4635db045cc2b861ef15b0#file-gistfile1-txt

**Comment by Endmateria** [comment]
I think the "java.util.ConcurrentModificationException" is someone else's mods conflicting, cuz I was still getting that error after removing your mod... so I would ignore that one.  Also, I guess the memory issue with pregenerators is a common problem... not singular to your mod.  Maybe if the step size was smaller, it wouldn't crash so much, I don't know, I'm not an expert on the subject.
